### ISSUE TITLE
get_hash function("SHA-256") always returns null

### ISSUE THREAD

**Issue by L1onKing** [randombit/botan#3063]
Hello everyone! I am having an issue with Botan, in particular that the function get_hash with parameter "SHA-256" returns null. I have checked BOTAN_HAS_SHA2_32 macro, it is defined.

I already pulled from the repo and built a botan 2.0 from the scratch (before I was using the library that was built by my colleague), and I still have this issue.

I tried to Google it, but it doesn't seem to be a common issue. Please guide me into the right direciton, how to solve this issue.

Thank you

**Comment by lieser** [comment]
Could you please post the code on how exactly you are calling the `get_hash()` function?

And please not that `get_hash()` is a deprecated function that was removed in newer versions. I would recommend to start using `Botan::HashFunction::create("SHA-256")` instead.
See also https://botan.randombit.net/handbook/api_ref/hash.html#code-example for a code example.

**Comment by L1onKing** [comment]
Hello @lieser ! Thank you for your response. Here is a piece of code that return null pointers:

```
std::unique_ptr<Botan::HashFunction> hash1(Botan::HashFunction::create("SHA-256"));
HashFunction* hash = get_hash("SHA-256");
```

In fact, I tried to use new API, but no success, I still get `null` instead of hash functions. I also tried to set `provider` argument to `"base"`, but it didn't help

**Comment by randombit** [comment]
Can you post how you built the library? The configure.py invocation is included in `botan/build.h` in the comment near the top of the file.

**Comment by L1onKing** [comment]
Hello @randombit ! Thank you for your response. Here is the piece of info you have asked:

```
/*
* Build configuration for Botan 2.19.2
*
* Automatically generated from
* 'configure.py --cc=msvc --os=windows --debug-mode'
*
* Target
*  - Compiler: cl  /EHs /GR /MDd /bigobj /Zi /FS
*  - Arch: x86_64
*  - OS: windows
*/
```

**Comment by lieser** [comment]
I don't see anything wrong with your call to `configure.py`.
But can you still check that `sha2_32` is listed in the *Loading modules* info then you call `configure.py`? Best also post the complete output from it.

I would suggest:
- make sure you are linking to the botan lib you build yourself
- try running the tests (`nmake check`) and/or CLI tool of botan (https://botan.randombit.net/handbook/cli.html#hash-function) and see if they have the same problem
- use a debugger and step into the `HashFunction::create()` function of Botan. The function is pretty straight forward and you should be able to see what is happening without much knowledge of Botan itself.

**Comment by L1onKing** [comment]
@lieser 
`But can you still check that sha2_32 is listed in the Loading modules info then you call configure.py? Best also post the complete output from it.`

Could you please advice how to do that?

**Comment by L1onKing** [comment]
@lieser @randombit Ok guys, I figured out the problem. But now I have a few questions. Let me dive into the core of the problem first:

I managed to fix it by rebuilding my botan from `release-2` branch. The problem was that I linked to new headers and .lib file, but ignored `.dll` file. I was under the impression that I'm building static botan library, but apparently I have built a dynamic library where `.lib` is used just as a pointers space to functions stored in `.dll`

That got me confused a bit. Can Botan be built as a core static library, where there's just only one `.lib` file with all functions inside? Or is it possible to build only as a dynamic library? Is this logic (having `.lib` file as a wrapper for accessing functions in `.dll` library) a common practice with Botan or with Windows in general?

I am sorry if my questions might be stupid. Last 10 years I was using Mac OS and iOS, and I was using Botan last few years pretty much successfuly and without any issue on these platforms

**Comment by lieser** [comment]
On Windows the default for Botan is building it as a shared library. But you can also change it to static. 

Either use:
- `--enable-static-library`
- [`--disable-shared-library`](https://botan.randombit.net/handbook/building.html#disable-shared-library) (probably not that intuitive, but it will then default to static)
- or be explicit which targets you want to build with [--build-targets=](https://botan.randombit.net/handbook/building.html#build-targets-build-targets), e.g. `--build-targets=static,cli`  for building static library and the cli, but not the tests.

Note that on windows only either static or shared can be build at the same time. Not both.

**Comment by L1onKing** [comment]
@lieser Thank you for your response! Now it's much clear to me.

One thing though. I noticed that the library only works for me if I build it with the command:

`python configure.py --cc=msvc --os=windows --debug-mode`

So if I remove `--debug-mode`, I have the same issue again. Could you advise what the problem can be here? Thank you!

**Comment by reneme** [comment]
> Is this logic (having .lib file as a wrapper for accessing functions in .dll library) a common practice with Botan or with Windows in general?

Yes, that's how windows tackles dynamic linkage. Where on macOS you would simply see a single *.dylib file, Windows uses the *.lib and *.dll files basically exactly as you described.

That said: Could it be that you again have some mixed up library file? You might even mix up non-debug and debug libraries.
Please remove all *.lib and *.dll files from botan's build folder and try again.

**Comment by L1onKing** [comment]
Hello @reneme! Thank you for you comment, much appreciated. I will do that again and let you know. I think you're probably right, I might have fixed up something.

**Comment by L1onKing** [comment]
I double checked my integration process, and I am positive I'm linking to all files - new include, new `.lib` and new `.dll`. I still have this issue. If I build with `--debug-mode`, everything runs. I will continue investigating it and update the thread once I find the answer.

**Comment by randombit** [comment]
Closing, if you are still having a problem please open a new issue
